woocommerce/woocommerce-ios
# Integrate In-app Feedback Card Conditions for Displaying

Closes #2537. 

This depends on #2623 which has been reviewed and approved but hasn't been merged yet. I'll merge it after the release cut.

This PR integrates the action [`AppSettingsAction.loadInAppFeedbackCardVisibility`](https://github.com/woocommerce/woocommerce-ios/blob/fd208e19dee798df19578b636ba03a4a075dd5a7/Yosemite/Yosemite/Actions/AppSettingsAction.swift#L83-L90) (#2623) to the card UI. The UI will now only show up in these conditions:

1. Users who’ve installed the app at least 3 months ago.
2. Once every 6 months. (this will have to be tested later)

ℹ️  The buttons' functionalities will be integrated later with #2626 and #2605. 

## Changes 

### ViewModel

I moved the `canDisplayInAppFeedbackCard` logic into a new ViewModel, [`StoreStatsAndTopPerformersPeriodViewModel`](https://github.com/woocommerce/woocommerce-ios/blob/issue/2537-in-app-feedback-card-visibility-conditions/WooCommerce/Classes/ViewRelated/Dashboard/Stats%20v4/StoreStatsAndTopPerformersPeriodViewModel.swift). This ViewModel is responsible for dictating when the in-app feedback card should be shown. 

Later, in #2629, the buttons will also call a function inside here to call [`AppSettingsAction.setLastFeedbackDate`](https://github.com/woocommerce/woocommerce-ios/blob/f0828fe1f156c33d44b37614574f697dedbed35a/Yosemite/Yosemite/Actions/AppSettingsAction.swift#L78-L81). 

### UI

The UI uses the new `isInAppFeedbackCardVisible` `Observable` exposed by the ViewModel to hide or show the in-app feedback card: 

https://github.com/woocommerce/woocommerce-ios/blob/fd208e19dee798df19578b636ba03a4a075dd5a7/WooCommerce/Classes/ViewRelated/Dashboard/Stats%20v4/StoreStatsAndTopPerformersPeriodViewController.swift#L173-L190

## Testing

We can only easily test the condition number 1 for now: "Users who’ve installed the app at least 3 months ago.". The second condition will be tested later after the functionalities of the buttons are integrated.

1. Delete the app on the device. 
2. Run the app and log in. 
3. Time travel to at least 90 days from now. 
4. Navigate to the Orders tab and navigate back to My Store.
5. Confirm that the in-app feedback card is now visible.
6. Time travel back to the present. 
7. Navigate to the Orders tab and navigate back to My Store.
8. Confirm that the in-app feedback card is no longer visible.

## Submitter Checklist
